### ISSUE TITLE
[BUG]: ext-psr should be a requirement or psr packages should not be provided

### ISSUE THREAD

**Issue by stof** [phalcon/cphalcon#15194]
**Describe the bug**
Composer 2 now respect `provide` rule better than Composer 1. This means that if you both require a package that exists on Packagist (like `psr/container`) and a package that provides it (like `phalcon/cphalcon`).
This is causing issues for projects migrating to Composer 2. See https://github.com/composer/composer/issues/9382.

Currently, `phalcon/cphalcon` says it provides the psr interfaces. but they are actually not provided if `ext-psr` is not loaded in PHP, which causes issues.

I see 2 alternatives:
- add `ext-psr` as a requirement (rather than a dev requirement) to ensure it is always there
- remove the `provide` rules. This will actually not change anything for Composer 1 users as Composer 1 already keeps `psr/container` (and other psr packages) in the vendor folder as they are actual packages and not virtual packages.

**Comment by Jeckerson** [comment]
Hello @stof

PSR is required only for dev, because `cphalcon` is compiled PHP extension, which means you install it via `pecl` or OS package manager. So only during the development process, `composer install` is executed.

**Comment by stof** [comment]
@Jeckerson that's not true if a project adds `phalcon/cphalcon` as a requirement but does not install `ext-psr` (which is perfectly allowed currently, which is the whole point of this issue). In such case, Composer 2 won't install the `psr/*` packages (because of the provide rule) and the interfaces will be missing. See the original report in the composer repository (the link is above in the issue) for such a case.

**Comment by Jeckerson** [comment]
Let me rephrase it. There is no point to require `phalcon/cphalcon` project as not `--dev`, because all work occur inside dev environment. Only inside `tests/` folder are PHP files, all other files - are [Zephir (.zep) files](https://github.com/phalcon/zephir), which are compilable.

People who include `phalcon/cphalcon` inside their projects might not know how Phalcon framework works or have different goals or purposes. And currently they are only 16:

![image](https://user-images.githubusercontent.com/3289702/100651987-3380cc80-333e-11eb-89d8-dd0790a0830a.png)

**Comment by stof** [comment]
@Jeckerson even if they include it as `--dev`, they might have other tools in `require-dev` than need to be able to use PSR interfaces themselves (Behat for instance), and then suffer from the same issue.
